File: src/main/java/june_2025/MissingNumber.java
package june_2025;

public class MissingNumber {

    public static void main(String[] args) {
        int[] nums = {3, 0, 1};
        MissingNumber missingNumber = new MissingNumber();
        int result = missingNumber.missingNumber(nums);
        System.out.println(result);
    }

    // O(n) time | O(1) space
    public int missingNumber(int[] nums) {
        int n = nums.length;
        int xor = n;

        for (int i = 0; i < nums.length; i++) {
            xor ^= i ^ nums[i];
        }
        return xor;
    }

}

File: python/missing_number.py
class Solution(object):
    # O(n) time | O(1) space
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res = len(nums)

        for i in range(len(nums)):
            res += i - nums[i]
        return res


solution = Solution()
print(solution.missingNumber([9,6,4,2,3,5,7,0,1]))